src/main/BlockPool.cpp

```diff
@@ -0,0 +1,1216 @@
+
+// NUM_SMALL_POOLS = (256 / 8) + 1 = 33
+#define NUM_SMALL_POOLS (33)
+
+// Constant indicating a deleted bucket.
+#define EMPTY_KEY (0)
+#define DELETED_KEY (UINTPTR_MAX)
+
+// Samll object arena is 1Gb = 30 bits.
+#define ARENASIZE (1 << 30)
+
+/*
+ * A small object superblock has size 2^18.
+ * Minimum object size = 32 bytes.
+ */
+#define SUPERBLOCK_BITS (18)
+#define SUPERBLOCK_SIZE (1 << SUPERBLOCK_BITS)
+
+/*
+ * Medium object superblocks use 2^19 bytes.
+ * Minimum object size = 64 bytes.
+ */
+#define MEDIUM_SUPERBLOCK_SIZE_LOG2 (19)
+#define MEDIUM_SUPERBLOCK_SIZE (1 << MEDIUM_SUPERBLOCK_SIZE_LOG2)
+
+/*
+ * NOTE: We use a fixed superblock header size, regardless of block size. This
+ * leaves some unused bitset entries for larger block sizes.
+ *
+ * The fixed header size is 1168 bytes. This is the breakdown:
+ *    block_size     =   32 bytes
+ *    next_untouched =   32 bytes
+ *    free_list      =   64 bytes
+ *    free bitset    = 1040 bytes
+ *    total          = 1168 bytes
+ *
+ * The bitset can map 1040 * 8 = 8320 blocks.  The minimum block size for a
+ * 2^18 superblock is 32 bytes, meaning the bitset covers 8320 * 8 = 266240
+ * bytes, which is more than the required range 2^18 - 1168 = 260976.
+ *
+ * For a 2^19 size superblock the minimum object size is 64 bytes,
+ * and 1040 * 8 * 64 = 532480 < 2^19 - 1168. So all objects are mappable
+ * for a 2^19 superblock.
+ */
+#define SUPERBLOCK_HEADER_SIZE (1168)
+
+typedef std::uint64_t u64;
+typedef std::uint32_t u32;
+
+using std::function;
+using std::size_t;
+
+#ifdef ALLOCATION_CHECK
+typedef std::map<void*, void*> allocation_map;
+static allocation_map allocations;
+
+static void add_to_allocation_map(void* allocation, size_t size);
+static void remove_from_allocation_map(void* allocation);
+static void* lookup_in_allocation_map(void* tentative_pointer);
+#endif
+
+/* Function to report allocation errors. */
+static void allocerr(const char* message) {
+  fprintf(stderr, "ERROR: %s\n", message);
+  abort();
+}
+
+#ifdef HAVE_ADDRESS_SANITIZER
+
+/*
+ * When compiling with the address sanitizer, we modify the allocation routines
+ * to detect errors more reliably.  In particular:
+ * - A redzone is added on both sides of the object to detect underflows and
+ *   overflows.
+ * - Memory is poisoned when it is freed to detect use-after-free errors.
+ * - Freed objects are held in a quarantine to prevent the memory from
+ *   being reallocated quickly.  This helps detect use-after-free errors.
+ *
+ * Note that in the context of GC, 'use after free' is really premature garbage
+ * collection.
+ */
+
+// Maximum quarantine size = 10Mb.
+#define MAX_QUARANTINE_SIZE (10<<20)
+
+// Redzones are added before and after the allocation.
+#define REDZONE_SIZE (16)
+
+// Quarantines are used so objects are not quickly reused after free.
+static size_t small_quarantine_size = 0;
+static size_t quarantine_size = 0;
+static FreeListNode* small_quarantine[NUM_SMALL_POOLS];
```

Comment:
This won't compile with HAVE_ADDRESS_SANITIZER, as FreeListNode hasn't been declared yet.